controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -30,32 +30,52 @@ import (
 
 // etcdClientGenerator generates etcd clients that connect to specific etcd members on particular control plane nodes.
 type etcdClientGenerator struct {
-	restConfig *rest.Config
-	tlsConfig  *tls.Config
+	restConfig   *rest.Config
+	tlsConfig    *tls.Config
+	createClient clientCreator
 }
 
-func (c *etcdClientGenerator) forNodes(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
-	endpoints := make([]string, len(nodeNames))
-	for i, name := range nodeNames {
-		endpoints[i] = staticPodName("etcd", name)
+type clientCreator func(ctx context.Context, endpoints []string) (*etcd.Client, error)
+
+// NewEtcdClientGenerator returns a new etcdClientGenerator instance.
+func NewEtcdClientGenerator(restConfig *rest.Config, tlsConfig *tls.Config) *etcdClientGenerator {
+	ecg := &etcdClientGenerator{restConfig: restConfig, tlsConfig: tlsConfig}
+
+	ecg.createClient = func(ctx context.Context, endpoints []string) (*etcd.Client, error) {
+		p := proxy.Proxy{
+			Kind:       "pods",
+			Namespace:  metav1.NamespaceSystem,
+			KubeConfig: ecg.restConfig,
+			TLSConfig:  ecg.tlsConfig,
+			Port:       2379,
+		}
+		return etcd.NewClient(ctx, endpoints, p, ecg.tlsConfig)
 	}
 
-	p := proxy.Proxy{
-		Kind:       "pods",
-		Namespace:  metav1.NamespaceSystem,
-		KubeConfig: c.restConfig,
-		TLSConfig:  c.tlsConfig,
-		Port:       2379,
+	return ecg
+}
+
+// forFirstAvailableNode takes a list of nodes and returns a client for the first one that connects.
+func (c *etcdClientGenerator) forFirstAvailableNode(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
+	var errs []error
+	for _, name := range nodeNames {
+		endpoints := []string{staticPodName("etcd", name)}
+		client, err := c.createClient(ctx, endpoints)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		return client, nil
 	}
-	return etcd.NewClient(ctx, endpoints, p, c.tlsConfig)
+	return nil, errors.Wrap(kerrors.NewAggregate(errs), "could not establish a connection to any etcd node")
```

Comment:
nit, add the list of nodeNames to the error message

Reply:
@shysank do you think you can add this node names soon?

Reply:
Actually let's get this in & we can adjust the error message later if needed, considering this has been on hold a while & Fabrizio's request is non-blocking.